semantic/poly_types.go

```diff
@@ -0,0 +1,778 @@
+package semantic
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+)
+
+// TypeExpression represents an expression describing a type.
+type TypeExpression interface {
+	// MonoType produces a monotype of the expression.
+	MonoType() (Type, bool)
+	// freeVars reports the free unbound type variables in the expression.
+	freeVars(*Constraints) TvarSet
+	// resolveType produces a monotype of the type expression given the kind constraints.
+	resolveType(map[Tvar]Kind) (Type, error)
+	// resolvePolyType applies the kind constraints producing a new self describing poly type.
+	resolvePolyType(map[Tvar]Kind) (PolyType, error)
+}
+
+//PolyType represents a polymorphic type, meaning that the type may have multiple free type variables.
+type PolyType interface {
+	TypeExpression
+	// occurs reports whether tv is a free variable in the type.
+	occurs(tv Tvar) bool
+	// substituteType replaces tv for t producing a new type.
+	substituteType(tv Tvar, t PolyType) PolyType
+	// unifyType unifies the two types given the kind constraints and produces a substitution.
+	unifyType(map[Tvar]Kind, PolyType) (Substitution, error)
+
+	// Equal reports if two types are the same.
+	Equal(PolyType) bool
+
+	// Nature reports the primitive description of the type.
+	Nature() Nature
+}
+
+// Kind is a constraint in the kind domain.
+type Kind interface {
+	TypeExpression
+	// substituteKind replaces occurences of tv with t producing a new kind.
+	substituteKind(tv Tvar, t PolyType) Kind
+	// unifyKind unifies the two kinds producing a new merged kind and a substitution.
+	unifyKind(map[Tvar]Kind, Kind) (Kind, Substitution, error)
+}
+
+// Tvar represents a type variable meaning its type could be any possible type.
+type Tvar int
+
+func (tv Tvar) Nature() Nature {
+	return Invalid
+}
+func (tv Tvar) String() string {
+	if tv == 0 {
+		// tv == 0 is not considered valid,
+		// we denote that by using a different
+		// symbol other than t0.
+		return "><"
+	}
+	return fmt.Sprintf("t%d", int(tv))
+}
+
+func (a Tvar) occurs(b Tvar) bool {
+	return a == b
+}
+func (a Tvar) substituteType(b Tvar, t PolyType) PolyType {
+	if a == b {
+		return t
+	}
+	return a
+}
+func (tv Tvar) freeVars(c *Constraints) TvarSet {
+	fvs := TvarSet{tv}
+	if c != nil {
+		ks, ok := c.kindConst[tv]
+		if ok {
+			for _, k := range ks {
+				fvs = fvs.union(k.freeVars(c))
+			}
+		}
+	}
+	return fvs
+}
+func (l Tvar) unifyType(kinds map[Tvar]Kind, r PolyType) (Substitution, error) {
+	switch r := r.(type) {
+	case Tvar:
+		if l == r {
+			return nil, nil
+		}
```

Comment:
Maybe this happens somewhere else, but it looks like it's missing an occurs check. How does `(x) => x(x:x)` type check?

Reply:
The occurs check is here https://github.com/influxdata/flux/pull/203/files#diff-594e3760d97b20e5e28945e9d3c70f19R165

That if condition is not about occurs rather it  ignores any unification of a type variable with itself.

Reply:
Of course the occurs check doesn't matter when unifying two variables. 👍